src/serializer/brat.py

```diff
@@ -0,0 +1,250 @@
+import json
+import os
+from collections import defaultdict
+from typing import DefaultDict, Dict, List, Optional, Sequence, Tuple, TypeVar
+
+from pie_datasets.core.dataset_dict import METADATA_FILE_NAME
+from pie_modules.annotations import BinaryRelation, LabeledMultiSpan, LabeledSpan
+from pie_modules.documents import TextBasedDocument
+from pytorch_ie import AnnotationLayer
+from pytorch_ie.core import Document
+from pytorch_ie.core.document import Annotation, BaseAnnotationList
+from pytorch_ie.utils.hydra import serialize_document_type
+
+from src.serializer.interface import DocumentSerializer
+from src.utils import get_pylogger
+
+log = get_pylogger(__name__)
+
+D = TypeVar("D", bound=Document)
+
+
+def serialize_labeled_spans(
+    entities: BaseAnnotationList,
+    label_prefix: Optional[str] = None,
+    first_span_id: int = 0,
+) -> Tuple[List[str], Dict[LabeledSpan, str]]:
+    """Converts entity annotations of type LabeledMultiSpan and LabeledSpan to annotations in the
+    Brat format.
+
+    Parameters:
+        entities (BaseAnnotationList): The list of entity annotations.
+        label_prefix (Optional[str]): An optional prefix to add to entity labels.
+        first_span_id: An integer value used for creating span annotation IDs. It ensures the proper assignment of IDs
+            for predicted annotations, particularly when gold annotations have already been included.
+
+    Returns:
+        Tuple[List[str], DefaultDict[LabeledSpan, str]]: A tuple containing a list of strings representing
+        entity annotations in the Brat format, and a dictionary mapping labeled spans to their IDs.
+    """
+    span2id: Dict[LabeledSpan, str] = defaultdict()
+    ann_entities = []
+    for i, entity in enumerate(entities, start=first_span_id):
+        entity_id = f"T{i}"
+        label = entity.label if label_prefix is None else f"{label_prefix}-{entity.label}"
+        if isinstance(entity, LabeledMultiSpan):
+            locations = []
+            texts = []
+            for slice in entity.slices:
+                start_idx = slice[0]
+                end_idx = slice[1]
+                texts.append(entity.target[start_idx:end_idx])
+                locations.append(f"{start_idx} {end_idx}")
+            location = ";".join(locations)
+            text = " ".join(texts)
+            entry = f"{entity_id}\t{label} {location}\t{text}\n"
+        elif isinstance(entity, LabeledSpan):
+            start_idx = entity.start
+            end_idx = entity.end
+            entity_text = entity.target[start_idx:end_idx]
+            entry = f"{entity_id}\t{label} {start_idx} {end_idx}\t{entity_text}\n"
+        else:
+            raise Warning(f"labeled span has unknown type: {type(entity)}")
+        span2id[entity] = entity_id
+        ann_entities.append(entry)
+    return ann_entities, span2id
+
+
+def serialize_binary_relations(
```

Comment:
As above, can you outsource some code to a method `serialize_binary_relation(idx: int, labeled_span: Union[LabeledMultiSpan, LabeledSpan], span2id: Dict[Annotation, str], label_prefix: Optional[str] = None) -> str`?